### ISSUE TITLE
go tool vet called incorrectly

### ISSUE THREAD

**Issue by rbradford** [alecthomas/gometalinter#307]
After #289 "go tool vet" is now called with a list of all the files in the source tree being linted. However when go tool vet is called with a set of files it expects all the files to be in the same package:

See https://golang.org/cmd/vet/

> By files:
> 
> go tool vet source/directory/*.go
> 
> vets the files named, all of which must be in the same package.

We're observing that this is leading to lint failures with example_test.go files using names defined in the same package.

The solution is to use "go tool vet" with the set of directories.

**Comment by teknico** [comment]
This was not fixed by #315, unfortunately:
```
$ go tool vet ./bloom/
$ gometalinter --disable-all --enable=vet --vendor --debug ./bloom/
[...]
DEBUG: linting path ./bloom
DEBUG: [vet.1]: executing ~/bin/go tool vet bloom/example_test.go bloom/filter.go \
  bloom/filter_test.go bloom/merkleblock.go bloom/merkleblock_test.go \
  bloom/murmurhash3.go bloom/murmurhash3_test.go
DEBUG: [vet.1]: warning: ~/bin/go returned exit status 1: bloom/example_test.go:20: \
  ExampleNewFilter refers to unknown identifier: NewFilter
[...]
```
Using `master` at commit `bfcc1d6942136fd86eb6f1a6fb328de8398fbd80` on the `github.com/decred/dcrutil` repo at commit `ddbde93f65ab0692e54ed8a5ad325fa2e8af4daa` with Go 1.9.

**Comment by dnephin** [comment]
I believe this is only a problem when there are multiple packages in the same directory path (ex: `bloom` and `bloom_test`). I've run into this problem with an `example_test.go` as well.  I don't think this is a problem with #315, I believe the code worked the same way before that change (calling `go tool vet`  with all the `.go` files in a directory.

The workaround I used was to add `example_test.go` to the exclude list. https://github.com/gotestyourself/gotestyourself/blob/master/gometalinter.json#L5

If you call `go tool dev ./bloom/` does it actually catch vet errors in the `example_test.go` or is effectively equivalent to excluding the file?  I think the problem with using the directory path is that `go tool vet` will recursively vet everything under that path, which is not the behaviour we want.

**Comment by teknico** [comment]
> I believe this is only a problem when there are multiple packages in the same directory path (ex: bloom and bloom_test). I've run into this problem with an example_test.go as well.

Indeed, that's what's going on.

> I don't think this is a problem with #315,

#315's description does mention fixing this bug, among others.

> I believe the code worked the same way before that change (calling go tool vet with all the .go files in a directory).

It did work the same way before, and that prevents `go vet` from working properly, in this case.

> The workaround I used was to add example_test.go to the exclude list. https://github.com/gotestyourself/gotestyourself/blob/master/gometalinter.json#L5

Well, I'd rather keep spurious warning than exclude code from checks.

> If you call `go tool dev ./bloom/`, does it actually catch vet errors in the example_test.go or is effectively equivalent to excluding the file?

`go vet` finds no actual vet errors in those test files, likely because there are none. :-) I see no reason to believe that it is excluding the test files.

> I think the problem with using the directory path is that go tool vet will recursively vet everything under that path, which is not the behaviour we want.

I'm not sure why you don't want that behavior, but ok. In any case, if the changes in #315 were meant to also fix this, something went wrong. :-)

**Comment by dnephin** [comment]
#315 was not meant to fix this. The problem you're encountering is not related to this issue, it's a limitation with how go vet accepts args (vets the files named, all of which must be in the same package).

**Comment by teknico** [comment]
> #315 was not meant to fix this.

Then maybe its description could be changed accordingly.

> The problem you're encountering is not related to this issue,

My problem seems to match pretty well with the description of this issue.

> it's a limitation with how go vet accepts args (vets the files named, all of which must be in the same package).

Given that this limitation exists, as mentioned in the `go vet` documentation, maybe `gometalinter` could take it into account and call `go vet` in a way that does not make it produce spurious output in common situations.

**Comment by dnephin** [comment]
> Then maybe its description could be changed accordingly.

Let me rephrase. #315 was not meant to fix the problem you have. It was meant to fix this issue (#307). They are different. This issue about about vet failing to lint a directory with a single package. Your issue is that it fails to lint a directory with multiple packages.

> maybe gometalinter could take it into account and call go vet in a way that does not make it produce spurious output in common situations.

The current behaviour is correct for the common case. Multiple packages per directory is not the common case.

I've opened #351 for this issue, so let's continue the discussion there, since the problem is not the same as this issue (#307).

**Comment by teknico** [comment]
> This issue about about vet failing to lint a directory with a single package. Your issue is that it fails to lint a directory with multiple packages.

You're right, sorry. I somehow missed the "using names defined in the same package" part. Moving to #351.
